### ISSUE TITLE
CDCC is a non-refundable credit

### ISSUE THREAD

**Issue by martinholmer** [PolicyEngine/openfisca-us#842]
After upgrading to the 0.68.1 version, I still get some test failures caused by the current `cdcc` formula lacking any non-refundable logic.  By definition,  the amount of a non-refundable credit can never be more than the pre-credit income tax liability.

To illustrate this problem, here is a test case in which a $1750 CDCC is assigned to a parent with zero taxable income:
```
(openfisca-us) OpenFisca-US% cat f19-id217.test.yaml 
# OpenFisca-US test file derived from the f19.its.csv and f19.ots.csv files
# Expected test output from NBER TAXSIM Model v35 (05/18/22) With TCJA

- name: Tax unit with taxsimid 217 from f19.its.csv
  absolute_error_margin: 0.01
  period: 2019
  input:
    people:
      person1:
        is_tax_unit_head: 1
        is_tax_unit_spouse: 0
        is_tax_unit_dependent: 0
        age: 66
        employment_income: 6000
        unemployment_compensation: 4000
        social_security: 21000
        pension_income: 2000
        has_marketplace_health_coverage: false
      person2:
        is_tax_unit_head: 0
        is_tax_unit_spouse: 0
        is_tax_unit_dependent: 1
        age: 12
        has_marketplace_health_coverage: false
      person3:
        is_tax_unit_head: 0
        is_tax_unit_spouse: 0
        is_tax_unit_dependent: 1
        age: 12
        has_marketplace_health_coverage: false
      person4:
        is_tax_unit_head: 0
        is_tax_unit_spouse: 0
        is_tax_unit_dependent: 1
        age: 16
        has_marketplace_health_coverage: false
      person5:
        is_tax_unit_head: 0
        is_tax_unit_spouse: 0
        is_tax_unit_dependent: 1
        age: 20
        has_marketplace_health_coverage: false
    tax_units:
      tax_unit:
        members: [person1, person2, person3, person4, person5]
        tax_unit_childcare_expenses: 5000
    households:
      household:
        members: [person1, person2, person3, person4, person5]
        state_code: TX
  output:
    tax_unit_taxable_unemployment_compensation: 4000.00
    tax_unit_taxable_social_security: 0.00
    adjusted_gross_income: 12000.00
    taxable_income: 0.00
    cdcc: 0.00
    non_refundable_ctc: 0.00
    refundable_ctc: 525.00
    eitc: 2700.00
    income_tax: -3225.00

(openfisca-us) OpenFisca-US% openfisca-us test f19-id217.test.yaml 
---[snip]---
  Test 'Tax unit with taxsimid 217 from f19.its.csv':
    cdcc@2019: [1750.] differs from 0.0 with an absolute margin [1750.] > 0.01
---[snip]---
====================== 1 failed in 0.06s ========================
```

@MaxGhenis @nikhilwoodruff

**Comment by nikhilwoodruff** [comment]
Yes, so I think this is more a definitional issue than an error: [we cap the aggregate of the non-refundable credits](https://github.com/PolicyEngine/openfisca-us/blob/master/openfisca_us/variables/gov/irs/credits/non_refundable.py), rather than each on in turn, which would require us to add some assumptions about the order each one is capped in. I think this is closer to [the law](https://www.law.cornell.edu/uscode/text/26/26#a), which in the linked section also caps the aggregate, rather than each credit. What do you think is best here?

**Comment by martinholmer** [comment]
@nikhilwoodruff said in issue #842:

> Yes, so I think this is more a definitional issue than an error: [we cap the aggregate of the non-refundable credits](https://github.com/PolicyEngine/openfisca-us/blob/master/openfisca_us/variables/gov/irs/credits/non_refundable.py), rather than each in turn, which would require us to add some assumptions about the order each one is capped in. I think this is closer to [the law](https://www.law.cornell.edu/uscode/text/26/26#a), which in the linked section also caps the aggregate, rather than each credit. What do you think is best here?

Thanks for point this out.  I had not noticed the aggregate credit capping logic.

The test case in issue #842 does have the same income tax liability as TAXSIM35 gets, so OpenFisca-US is doing the tax calculations correctly for this case.

I find the legal language in the "law" incomprehensible so I don't know if aggregate credit capping "is closer to the law" than sequential capping.

What I do know is that the IRS 1040 forms do present the taxpayer with sequential capping and that both the TAXSIM35 and Tax-Calculator models implement sequential capping.  Both aggregate capping and sequential capping (if correctly done) will produce the same income tax liability.   The use of a different capping logic just makes it slightly more difficult to find the cause of income tax differences between OpenFisca-US and TAXSIM35.

So, I'll close this issue now.   Sorry for the false alarm.

@MaxGhenis

**Comment by nikhilwoodruff** [comment]
Thanks @martinholmer.

> I find the legal language in the "law" incomprehensible so I don't know if aggregate credit capping "is closer to the law" than sequential capping.

If it helps, here's my understanding of the law:

> The aggregate amount of credits allowed by this subpart **("this subpart" = Subpart A, containing all non-refundable credit definitions)** for the taxable year shall not exceed the sum of [income tax + alternative minimum tax]

I don't think the law even specifies how to limit, and just leaves it to the IRS to decide. I have no real opposition to using whichever sequential logic other models, and the IRS use. But thanks for this issue and discussion here.

**Comment by martinholmer** [comment]
@nikhilwoodruff said in closed issue #842:
> I don't think the law even specifies how to limit, and just leaves it to the IRS to decide. I have no real opposition to using whichever sequential logic other models, and the IRS use.

The benefits of switching to sequential credit capping are probably less than the non-trivial costs of switching.  Perhaps make no changes now and monitor this issue going forward.

@MaxGhenis
